### DISCUSSION TITLE
Multiline match mode for regex managers

### DISCUSSION THREAD

**Discussion by satwell** [renovatebot/renovate#30170] [Suggest an Idea]
### Tell us more.

Would it be possible for regex manager to use multi-line mode (the [`m` flag](https://github.com/google/re2/wiki/Syntax#Flags) for regular expressions) when applying the regular expressions in `matchStrings`?  This would let users use `^` and `$` to anchor to lines instead of the whole file.

It's handy for many syntaxes to match on line boundaries.  E.g., to avoid matches in line-level comments.  It's possible to get the same behavior by using `(?:^|\r\n|\r|\n)` to match line beginnings and `(?:$|\r\n|\r|\n)` to match line ends.  But this makes the regexes much messier.

I think the behavior of `^` and `$` only matching file beginning and end can trip up new users when trying to figure out this already complex feature ([example](https://github.com/renovatebot/renovate/discussions/27256)).  It certainly tripped me up, even though I'd read the [docs](https://docs.renovatebot.com/modules/manager/regex/#regular-expression-capture-groups) where they say:

> The regex manager matches are done per-file and not per-line, you should be aware when using the `^` and/or `$` regex assertions.

I initially interpreted this as "we use multi-line mode so be aware that `^` and `$` will match line boundaries within the file" because it seems like obviously desirable behavior, although I realize now that this was incorrect.

**Comment by satwell** [comment]
In the short term, I think clarification in the [docs](https://docs.renovatebot.com/modules/manager/regex/#regular-expression-capture-groups) would be helpful.  Instead of just telling users to "be aware when using `^` and/or `$`," I'd suggest saying something like this:

> The `regex` manager matches are done per-file and not per-line. This means the `^` and `$` regex assertions only match the beginning and end of the entire file. You can use `(?:^|\r\n|\r|\n|$)` if you need to match line boundaries.

I was about to submit a PR for the docs, but then realized I'd have to sign the CLA, which is not permitted by my employment contract.

**Reply by viceice** [reply]
provide it with a private account?

**Reply by risu729** [reply]
I also faced this issue, so opened a [PR](https://github.com/renovatebot/renovate/pull/34571). Thanks for the workaround!
